### PR TITLE
[14.0][MIG] sale order restrict cancel existing invoice

### PR DESCRIPTION
**Pull request by xavier-bouquiaux**
Hi All,

In v14, invoice.state = open became move_type = out_invoice

Kr

Xbo.

@rousseldenis

### CODE REVIEW

**Comment by rousseldenis** [comment]
> Hi All,
> 
> In v14, invoice.state = open became move_type = out_invoiceo
> 
> Kr
> 
> Xbo.
> 
> @rousseldenis

@xavier-bouquiaux Not exactly.

move_type is the field to distinguish pure account moves and invoices.

state field is still there but disappeared states for payment have been replace by payment_state field.

https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_move.py#L139

https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_move.py#L225

File: sale_order_restrict_cancel_existing_invoice/models/sale_order.py [hunk 1]
```diff
@@ -10,24 +10,24 @@ class SaleOrder(models.Model):
...
     )
 
     def _get_confirmed_invoice_states(self):
         """
-        Return the invoice states that should avoid order cancel
+        Return the invoice move_type that should avoid order cancel
         :return:
         """
-        return ["open"]
+        return ["out_invoice"]
```
Comment (rousseldenis) [inline sale_order_restrict_cancel_existing_invoice/models/sale_order.py]: Should be 'posted'

File: sale_order_restrict_cancel_existing_invoice/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    def action_cancel(self):
+        if any(sale.confirmed_invoice_ids for sale in self):
+            raise UserError(
+                _(
+                    "You cannot cancel a Sale Order for which a confirmed "
+                    "invoice exists!"
+                )
+            )
+        return super(SaleOrder, self).action_cancel()
```
Comment (rousseldenis) [inline sale_order_restrict_cancel_existing_invoice/models/sale_order.py:41]: @xavier-bouquiaux Could you add a test that check that normal cancelling is still working ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Please see comments

File: sale_order_restrict_cancel_existing_invoice/models/sale_order.py (hunk 1 above)
Comment (rousseldenis) [reply sale_order_restrict_cancel_existing_invoice/models/sale_order.py]: Ok, the function name is confusing.

Could you update it ?

`_get_confirmed_invoice_types`

Comment (Cedric-Pigeon) [reply sale_order_restrict_cancel_existing_invoice/models/sale_order.py]: done

File: sale_order_restrict_cancel_existing_invoice/models/sale_order.py (hunk 2 above)
Comment (Cedric-Pigeon) [reply sale_order_restrict_cancel_existing_invoice/models/sale_order.py:41]: done

**Comment by Cedric-Pigeon** [comment]
> Hi All,
> 
> In v14, invoice.state = open became move_type = out_invoice
> 
> Kr
> 
> Xbo.
> 
> @rousseldenis

The check must be done on payment_state and not on move type. I just arranged.

**Comment by rousseldenis** [comment]
@Cedric-Pigeon @xavier-bouquiaux Could you check travis and pre-commit ?

**Pull request closed without merging**